File: 01. Regresion lineal simple.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jul 29 21:00:08 2021

@author: Admin
"""

### Algoritmo de regresión lineal simple en py ###

# =============================================================================
# libraries
# =============================================================================

import pandas as pd
#import numpy as np
import matplotlib.pyplot as plt
import sklearn.model_selection as sklm
import sklearn.linear_model as skllm

# =============================================================================
# read data
# =============================================================================

salarios = pd.read_csv("data\Salary_data.csv")

# =============================================================================
# Datos entrenamiento y validacion
# =============================================================================

x_salarios = salarios.iloc[:, :-1].values
y_salarios = salarios.iloc[:, 1].values

x_train, x_test, y_train, y_test = sklm.train_test_split(x_salarios, 
                                                         y_salarios, 
                                                         test_size = 1/3,
                                                         random_state = 0)

# =============================================================================
# Modelo regresión lineal simple
# =============================================================================

regresion = skllm.LinearRegression()
regresion.fit(x_train, y_train)

y_pred = regresion.predict(x_test)

# =============================================================================
# Visualización
# =============================================================================

plt.scatter(x_train, y_train, color = "darkred")
plt.plot(x_train, regresion.predict(x_train), color = "darkblue")
plt.title("Sueldo vs Años de experiencia \nEntrenamiento")
plt.xlabel("Años de experiencia")
plt.ylabel("Sueldo ($)")
plt.show()

# con datos entrenamiento
plt.scatter(x_test, y_test, color = "darkred")
plt.plot(x_train, regresion.predict(x_train), color = "darkblue")
plt.title("Sueldo vs Años de experiencia \nValidación")
plt.xlabel("Años de experiencia")
plt.ylabel("Sueldo ($)")
plt.show()

File: 01. Regresion lineal simple.R
## Regresión lineal simple ##

# libraries ---------------------------------------------------------------

library(tidyverse)
library(caTools)
library(broom)

# read data ---------------------------------------------------------------

salarios <- readr::read_delim("data/Salary_Data.csv") # usé read_delim para 
                        # probar el "Delimiter guessing" de readr 2.0.0
                        # caso contrario, read_csv() basta y sobra

# entrenamiento y validacion ----------------------------------------------

set.seed(123)

split <- caTools::sample.split(salarios$Salary, SplitRatio = 2/3)

entrenamiento <- salarios %>% 
  subset(split == T)

validacion <- salarios %>% 
  subset(split == F)


# regresion lineal simple -------------------------------------------------

regresion <- entrenamiento %>% 
  lm(Salary ~ ., data = .)

summary(regresion)

# Visualización de resultados ---------------------------------------------

regresion %>% 
  broom::augment() %>% # para probar el tidy
  ggplot2::ggplot(aes(x = YearsExperience, y = Salary)) +
  geom_point(color = "darkred") + 
  geom_line(aes(y = .fitted), color = "darkblue") +
  labs(title = "Sueldo vs Años de experiencia \n(Entrenamiento)",
       y = "Salario",
       x = "Años de experiencia") +
  theme_classic()

ggplot2::ggplot() +
  geom_point(aes(x = validacion$YearsExperience, 
                 y = validacion$Salary),
             color = "darkred") + 
  geom_line(aes(x = entrenamiento$YearsExperience,
                y = regresion$fitted.values), 
            color = "darkblue") +
  labs(title = "Sueldo vs Años de experiencia \n(Validacion)",
       y = "Salario",
       x = "Años de experiencia") +
  theme_classic()
